### PR TITLE
BR: support restore compacted logs with checkpoint

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/57535

Problem Summary:

To support checkpoint and make switch-mode more robust.
### What changed and how does it work?
This PR add a checkpoint logic for restore compacted logs and improve the code of switch mode. 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/checkpoint/storage.go [hunk 1]
```diff
@@ -48,8 +47,9 @@ type checkpointStorage interface {
 // 2. BR regards the metadata table as a file so that it is not empty if the table exists.
 // 3. BR regards the checkpoint table as a directory which is managed by metadata table.
 const (
-	LogRestoreCheckpointDatabaseName      string = "__TiDB_BR_Temporary_Log_Restore_Checkpoint"
-	SnapshotRestoreCheckpointDatabaseName string = "__TiDB_BR_Temporary_Snapshot_Restore_Checkpoint"
+	LogRestoreCheckpointDatabaseName       string = "__TiDB_BR_Temporary_Log_Restore_Checkpoint"
+	SnapshotRestoreCheckpointDatabaseName  string = "__TiDB_BR_Temporary_Snapshot_Restore_Checkpoint"
+	CompactedRestoreCheckpointDatabaseName string = "__TiDB_BR_Temporary_Compacted_Restore_Checkpoint"
```
Comment (YuJuncen) [inline br/pkg/checkpoint/storage.go:52]: As there are more source of SSTs than compacting SST in PiTR, perhaps name it "SST_in_PiTR_Restore_Chcekpoint"?

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: br/pkg/storage/helper.go [hunk 2]
```diff
@@ -48,22 +52,32 @@ func UnmarshalDir[T any](ctx context.Context, walkOpt *WalkOption, s ExternalSto
 	errCh := make(chan error, 1)
 	reader := func() {
 		defer close(ch)
+		pool := util.NewWorkerPool(128, "metadata")
+		eg, ectx := errgroup.WithContext(ctx)
 		err := s.WalkDir(ctx, walkOpt, func(path string, size int64) error {
-			metaBytes, err := s.ReadFile(ctx, path)
```
Comment (Leavrth) [inline br/pkg/storage/helper.go:52]: check whether `ectx` is canceled here because the error is occurred in `ApplyOnErrorGroup` won't stop `WalkDir`.

**Review by Leavrth — APPROVED** [review]

File: br/pkg/checkpoint/storage.go (hunk 1 above)
Comment (3pointer) [reply br/pkg/checkpoint/storage.go:52]: I feel like `CustomSSTRestoreCheckpoint`, WDYT?

**Pull request merged**